llvm/lib/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.cpp

```diff
@@ -59,32 +67,116 @@ void checkKernel(const Function *F, const CallGraphTy &CG) {
   checkKernelImpl(F, CG, CallStack);
 }
 
+void computeFunctionToKernelsMappingImpl(Function *Kernel, const Function *F,
+                                         const CallGraphTy &CG,
+                                         FuncToFuncMapTy &Mapping) {
+  CallGraphTy::const_iterator It = CG.find(F);
+  // It could be that the function itself is a leaf and doesn't call anything
+  if (It == CG.end())
+    return;
+
+  Mapping[F].insert(Kernel);
+
+  const SmallPtrSet<Value *, 8> &Callees = It->getSecond();
+  for (const Value *V : Callees)
+    if (auto *Callee = dyn_cast<Function>(V))
+      computeFunctionToKernelsMappingImpl(Kernel, Callee, CG, Mapping);
+}
+
+void computeFunctionToKernelsMapping(Function *Kernel, const CallGraphTy &CG,
+                                     FuncToFuncMapTy &Mapping) {
+  CallGraphTy::const_iterator It = CG.find(Kernel);
+  // It could be that kernel doesn't call anything
+  if (It == CG.end())
+    return;
+
+  const SmallPtrSet<Value *, 8> &Callees = It->getSecond();
+  for (const Value *V : Callees) {
+    auto *Callee = dyn_cast<Function>(V);
+    if (!Callee)
+      continue;
+    Mapping[Callee].insert(Kernel);
+    computeFunctionToKernelsMappingImpl(Kernel, Callee, CG, Mapping);
+  }
+
+  // For simplicity we also consider a kernel to be using itself
+  Mapping[Kernel].insert(Kernel);
+}
+
+void collectVTablesThatUseFunction(
+    const Value *V, SmallVectorImpl<const GlobalVariable *> &VTables) {
+  for (const auto *U : V->users()) {
+    // GlobalVariable is also a constant
+    if (const auto *GV = dyn_cast<GlobalVariable>(U))
+      VTables.push_back(GV);
```

Comment:
maybe this is showing my own misunderstanding, but are we guaranteed all `GlobalVariable` users of an indrectly callable function are vtables? Is there nothing the user could do that would generate the same IR?

Reply:
In core SYCL no global variables can used by kernels unless they are `const`. There is `device_global` extension, but I doubt that it allows to store function pointers.

I suppose that there could be some cases which I don't know about with existing (and future) extensions, or with some implementation aspects that will introduce globals holding function pointers other than vtables. But for now we can probably assume that there are no such cases.

Worst case scenario, we will do some extra bit of online linking, that's the only effect from those attributes. I will add a comment explaining the current assumption to the code.

Reply:
I keep forgetting we don't allow globals, thanks.

Reply:
Added clarifying comment in 4ac51d473b1bc189fa77aae11516185c876ac41c